packages/services/src/contents/index.ts

```diff
@@ -1080,6 +1115,26 @@ export class Drive implements Contents.IDrive {
     this._apiEndpoint = options.apiEndpoint ?? SERVICE_DRIVE_URL;
     this.serverSettings =
       options.serverSettings ?? ServerConnection.makeSettings();
+    const contentProviderRegistry = new ContentProviderRegistry();
+    const restContentProvider = new RestContentProvider();
```

Comment:
If I am not missing something, it looks like the `Drive` no longer needs `_apiEndpoint` private member and we can instead pass it down to `RestContentProvider`:

```suggestion
    this.serverSettings =
      options.serverSettings ?? ServerConnection.makeSettings();
    const contentProviderRegistry = new ContentProviderRegistry();
    const restContentProvider = new RestContentProvider({
      apiEndpoint: options.apiEndpoint ?? SERVICE_DRIVE_URL
    });
```

Reply:
hmm `_apiEndpoint` is still used in `_getUrl()` in a bunch of places?

Reply:
Indeed. In that case I still think we should forward the configured API endpoint to `RestContentProvider`, right?

```suggestion
    this._apiEndpoint = options.apiEndpoint ?? SERVICE_DRIVE_URL;
    this.serverSettings =
      options.serverSettings ?? ServerConnection.makeSettings();
    const contentProviderRegistry = new ContentProviderRegistry();
    const restContentProvider = new RestContentProvider({
	  apiEndpoint: this._apiEndpoint
	});
```

Reply:
Good point. Either that or access the endpoint from the content provider through `this.drive.apiEndpoint` (if we make it public), or even `this.drive.getUrl()`. This is the reason why I set the content provider factory's `drive`, and it passes it on to each content provider. But I don't have a strong opinion, we could have content providers not keep a reference to their drive, and pass everything when we instantiate them. What do you think?

Reply:
I would prefer passing everything down during instantiation - this leads to smaller API surface and more predictable state.

Reply:
See https://github.com/jupyterlab/jupyterlab/pull/16744/commits/d228eb1af630891a9dccd40ae9d615e97f0ffea8 for a version where the REST content provider reference its drive to get the configuration.

Reply:
I think it would be better not to do that the way as in d228eb1. A a couple of reasons:
- it forces us to extend public API of drive by making `getUrl` public (maybe not the worst thing)
- it forces you to case `this.drive` which is `IDrive` initially to a `Drive` in an unsafe manner (this is if an extension author wanted to reuse the default `ContentProviderRegistry` implementation - although it is currently not exported - they could see a runtime error if they used it with a different `IDrive` implementation).
- it requires that the content provider registry sets the `drive` on each provider making making the logic flow more complex than it needs to be. This is because to register a content provider you the extension now already has to have access to the drive, so it could just pass it to content provider constructor.

What do you think about either:

```ts
new RestContentProvider({
   drive: this
});
```

or passing the `apiEndpoint` as proposed earlier?

This would allow to remove `drive` property from `IContentProviderRegistry` and from implementation.

Reply:
Then the `RestContentProvider` could require that drive is either of:
- `Drive` type or
- `IDrive & {getUrl: (...args[string]) => string}` interface (which we could name "IRestDrive")

Reply:
Yes, I removed the drive reference in content providers in https://github.com/jupyterlab/jupyterlab/pull/16744/commits/0430081c60d5e8a1e13aea69a45fe2560a2f7235. The configuration is passed when instantiated.